### DISCUSSION TITLE
Removed Resource from Avo and broke app

### DISCUSSION THREAD

**Discussion by rctneil** [avo-hq/avo#549] [Help] [answered]
Hi,

I deleted one resource from Avo but now my first resource is failing to display.

I get the following error:

```
NoMethodError in Avo::People#index
Showing /Users/name/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/avo-1.17.1/app/views/avo/partials/_turbo_frame_wrap.html.erb where line #2 raised:

undefined method `resources_spouse_path' for #<ActionDispatch::Routing::RoutesProxy:0x00007fc249094a08>
Did you mean?  resources_person_path
               resources_people_path
```

I've removed everything I can see relating to that other resource but it still fails to load.   Any ideas?

**Comment by adrianthedev** [comment]
Hey,

Have you restarted the server?

**Comment by rctneil** [comment]
Yes. Several times. No change unfortunately!

Thanks,
Neil

**Reply by adrianthedev** [reply]
Covering our bases here:

- restarted server
- restarted spring (if you have it) `spring stop`
- removed the resource file (maybe `app/avo/resources/spouse_resource.rb`)
- removed the controller file (maybe `app/controllers/avo/spouses_controller.rb`)

**Reply by adrianthedev** [reply]
Just to clarify, Avo doesn't "cache" anything like that. It's something left in memory or a rouge file from those ones👆

**Reply by rctneil** [reply]
Just tried all those and still getting the error.

It's very strange!

**Reply by rctneil** [reply]
Is this due to the fact that "Spouse" is a "type" of "Person" using Single Table Inheritance?

Not sure how it can be but thought i'd mention it.

**Reply by adrianthedev** [reply]
In theory, if you removed the resource file it should not appear anywhere in the app. 
You may remove that STI and see if that helps.

**Reply by rctneil** [reply]
If I change Spouse to inherit from Active Record Base instead and remove the relationships from both sides than I get the following error instead:

`ActiveRecord::SubclassNotFound in Avo::PeopleController#index`

`Invalid single-table inheritance type: Spouse is not a subclass of Person`

**Comment by rctneil** [comment]
It just looks like Avo is trying to use a specific route that is not available.  This is confirmed by checking `rails routes`, that route does not show up under Avo.   Why is it looking for that route in the first place?

It does have a feel about it like somethings being cached, but as you've said, Avo doesn't cache stuff.

**Reply by adrianthedev** [reply]
Yeah. Avo basically looks for all the resource files and creates routes for all of them. If there's no resource file, then, there's no route.

**Reply by rctneil** [reply]
I'm very confused then.  The resource for Spouse has been deleted, therefore the routes are no longer created but it's still trying to use them?

Any suggestions as to what I could look into next?   I'm sort of running out of ideas as to where the issue could lie.

**Reply by adrianthedev** [reply]
Hmm... do you have any associations towards `Spouse` set in other resources?

**Reply by rctneil** [reply]
This is my Person resource:

```
class PersonResource < Avo::BaseResource
  self.title = :forenames
  self.includes = []

  # self.search_query = ->(params:) do
  #   scope.ransack(id_eq: params[:q], m: "or").result(distinct: false)
  # end

  field :fullname, as: :text, name: 'Name', link_to_resource: true, only_on: [:index, :show]
  field :type, as: :text, name: 'Type', only_on: [:index, :show]
  field :title, as: :text, name: 'Title', only_on: [:edit]
  field :forenames, as: :text, name: 'Forenames', only_on: [:edit]
  field :surname, as: :text, name: 'Surname', only_on: [:edit]
  field :suffix, as: :text, name: 'Suffix', only_on: [:edit]
  field :nickname, as: :text, name: 'Nickname', only_on: [:edit]
  field :birthdate, as: :date, name: 'Date of birth'
  field :deathdate, as: :date, name: 'Date of death'
  field :parent, as: :text, name: 'Parent', only_on: [:index, :show] do |model|
      model.parent.fullname if model.has_parent?
  end 
  field :hidden, as: :boolean, name: 'Hidden'
  field :highlighted, as: :boolean, name: 'Highlighted'
  field :children, as: :number, name: 'Children', only_on: [:index] do |model|
      model.children.count
  end
  field :parent_id, as: :select, only_on: [:edit], name: 'Parent', options: ->(model: m, resource: r, view: v, field: f) {
      Person.people.order(forenames: :asc).all.map { |p| [p.fullname, p.id] }.to_h
  }, placeholder: 'Choose parent'
  field :children, as: :children, only_on: [:show]
  field :married_to, as: :text, name: 'Spouse', only_on: [:show] do |model|
    model.married_to.fullname
  end
  field :marriagedate, as: :date, name: 'Marriage Date', only_on: [:show], visible: -> (resource:) { resource.model.type == 'Spouse' }
  field :divorced, as: :boolean, name: 'Divorced?', only_on: [:show], visible: -> (resource:) { resource.model.type == 'Spouse' }

  filter PersonFilter
end
```

It's the only resource I have now.

My two models are as follows:

Person:

```
class Person < ApplicationRecord
    has_many :spouses, foreign_key: "marriage_person_id"
	has_ancestry cache_depth: true

    scope :people, -> { where(type: 'Person') }
    scope :spouses, -> { where(type: 'Spouse') }
    scope :no_ancestry, -> { people.where(ancestry: nil )}
    scope :visible, -> { where(hidden: false) }

    before_validation :build_fullname_for_db

    def fullname(spanned = false)
        {
            'title': title,
            'forenames': forenames,
            'surname': surname,
            'suffix': suffix
        }.map { |k,v| spanned ? "<span class='#{k}'>#{v}</span>" : v }.join(' ').html_safe
    end

    def has_lifetime?
        if birthdate || deathdate
            true
        else
            false
        end
    end

    # def get_date_parts(parts_available)
    #     def opening_date

    # end

    def lifetime_full
        if birthdate && deathdate
            if birthdateparts?
                if birthdateparts == 'Y'
                    bd = birthdate.strftime('%Y')
                elsif birthdateparts == 'MY'
                    bd = birthdate.strftime('%B %Y')
                end
            else
                bd = birthdate.strftime("#{birthdate.day.ordinalize} %B %Y")
            end

            if deathdateparts?
                if deathdateparts == 'Y'
                    dd = deathdate.strftime('%Y')
                elsif deathdateparts == 'MY'
                    dd = deathdate.strftime('%B %Y')
                end
            else
                dd = deathdate.strftime("#{deathdate.day.ordinalize} %B %Y")
            end

            bd + " - " + dd
        elsif birthdate
            if birthdateparts?
                if birthdateparts == 'Y'
                    bd = birthdate.strftime('%Y')
                elsif birthdateparts == 'MY'
                    bd = birthdate.strftime('%B %Y')
                end
            else
                bd = birthdate.strftime("#{birthdate.day.ordinalize} %B %Y")
            end

            "Born " + bd
        elsif deathdate
            if deathdateparts?
                if deathdateparts == 'Y'
                    dd = deathdate.strftime('%Y')
                elsif deathdateparts == 'MY'
                    dd = deathdate.strftime('%B %Y')
                end
            else
                dd = deathdate.strftime("#{deathdate.day.ordinalize} %B %Y")
            end

            "Died " + dd
        end
    end

    def lifetime_years
        if birthdate && deathdate
            birthdate.strftime("%Y") + " - " + deathdate.strftime("%Y")
        elsif birthdate
            "Born " + birthdate.strftime("%Y")
        elsif deathdate
            "Died " + deathdate.strftime("%Y")
        end
    end

    def has_marriages?
        marriage_person_id?
    end

    def build_fullname_for_db
        update_attribute(:fullnamedb, fullname)
    end
end
```

```
class Spouse < Person
    belongs_to :person, foreign_key: "marriage_person_id"

    def full_marriage_date
        if marriagedate
            if marriagedateparts?
                if marriagedateparts == 'Y'
                    md = marriagedate.strftime('%Y')
                elsif marriagedateparts == 'MY'
                    md = marriagedate.strftime('%B %Y')
                end
            else
                md = marriagedate.strftime("#{marriagedate.day.ordinalize} %B %Y")
            end

            md
        end
    end

    def married_to
        Person.find(marriage_person_id)
    end
end
```

**Reply by adrianthedev** [reply]
Hmm... nothing stands out.
I would suggest rebooting your computer. This way we can totally rule out any type of memory caching.

**Reply by adrianthedev** [reply]
Hey, I'm curious, has this been resolved?

**Reply by rctneil** [reply]
Hi @adrianthedev    Apologies for not replying on here.   Another project has sort of taken my attention the last couple of days.  I hope to be able to get back to this sometime next week.

I did reboot my machine and really hope that solves the issue. I will 100% update you on here with how I get on and for further troubleshooting if it persists.

Many thanks again though!

**Reply by adrianthedev** [reply]
Oh, no problem. Thanks for the update!

**Reply by rctneil** [reply]
@adrianthedev Just been working on this and unfortunately this is still causing an issue.   An entire reboot + OS update + other cache clears and stuff haven't resolves it.

Any suggestions?

**Reply by adrianthedev** [reply]
Putting some more ideas here:

- Check if any of your resources have the spouse model in `self.includes`
- Check that you don't have any associations towards the spouse resource in other Avo resources (belongs_to, has_many, etc) (seems the most obvious).
- Try and fiddle with the STI model. see if something changes if you disable that one.

Also, what is the path that gives you this error? Any path, and `Index` view or `Show` view?

If nothing works, let's get on a call so I can have a closer look. I really don't know what could cause it. It's really weird.

**Comment by rctneil** [comment]
Thanks for those points.  I hope to test them all out tomorrow and if none solve it then we can absolutely get on a call and try to fix this once and for all!

Thankyou again for your assistance,  I'll keep you updated!

Neil

**Reply by rctneil** [reply]
Hi @adrianthedev It only appears to happen on the route "/admin/resources/people".  "/admin" works fine on its own.

I can't seem to notice anything that corresponds with points 1 and 2 and I did try 3 the other day but not 100% sure I removed everything. The app is quite reliant on the STI so it's tricky to remove entirely.

Maybe we could somehow set up a call to look over this?

What's your preferred option here?

Thanks again,
Neil

**Reply by adrianthedev** [reply]
Got it.
Can you set something up for early next week?
https://calendly.com/adrianthedev

Thanks

**Comment by rctneil** [comment]
No problem.  Booked for 1pm Monday.  Calendly has not sent the invite to me yet as it stated. Hoping it will come later on.

Thanks,
Neil

**Comment by adrianthedev** [comment, marked as answer]
Just for the record, we couldn't figure out what the issue was and we ended up adding the resource back and hiding it from the sidebar.

**Comment by adrianthedev** [comment]
We're fixing this with https://github.com/avo-hq/avo/pull/560
